nevergrad/parametrization/core.py

```diff
@@ -0,0 +1,497 @@
+import typing as t
+import numpy as np
+
+
+BP = t.TypeVar("BP", bound="BaseParameter")
+P = t.TypeVar("P", bound="Parameter")
+D = t.TypeVar("D", bound="Dict")
+
+
+class Descriptors(t.NamedTuple):
+    """Provides access to a set of descriptors for the parametrization
+    This can be used within optimizers.
+    """
+    deterministic: bool = True
+    continuous: bool = True
+
+
+class NotSupportedError(RuntimeError):
+    """This type of operation is not supported by the parameter.
+    """
+
+
+# pylint: disable=too-many-instance-attributes
+class BaseParameter:
+    """This provides the core functionality of a parameter, aka
+    value, subparameters, mutation, recombination
+    """
+
+    def __init__(self, **subparameters: t.Any) -> None:
+        self.uid = uuid.uuid4().hex
+        self.parents_uids: t.List[str] = []
+        self._subparameters = None if not subparameters else Dict(**subparameters)
+        self._dimension: t.Optional[int] = None
+
+    @property
+    def descriptors(self) -> Descriptors:
+        return Descriptors(deterministic=True, continuous=True)
+
+    @property
+    def value(self) -> t.Any:
+        raise NotImplementedError
+
+    @value.setter
+    def value(self, value: t.Any) -> t.Any:
+        raise NotImplementedError
+
+    def spawn_child(self: BP) -> BP:
+        """Creates a new which only shares the same random generator than its parent
+
+        Returns
+        -------
+        Parameter
+            a new instance of the same class, with same value/parameters/subparameters/...
+        """
+        raise NotImplementedError
+
+    @property
+    def subparameters(self) -> "Dict":
+        if self._subparameters is None:  # delayed instantiation to avoid infinte loop
+            assert self.__class__ != Dict, "subparameters of Parameters dict should never be called"
+            self._subparameters = Dict()
+        assert self._subparameters is not None
+        return self._subparameters
+
+    def _get_parameter_value(self, name: str) -> t.Any:
+        param = self.subparameters[name]
+        return param.value if isinstance(param, Parameter) else param
+
+    @property
+    def dimension(self) -> int:
+        """Dimension of the standardized space for this parameter
+        i.e size of the vector returned by get_std_data()
+        """
+        if self._dimension is None:
+            try:
+                self._dimension = self.get_std_data().size
+            except NotSupportedError:
+                self._dimension = 0
+        return self._dimension
+
+    def mutate(self) -> None:
+        """Mutate subparameters of the instance, and then its value
+        """
+        self.subparameters.mutate()
+        data = self.get_std_data()  # pylint: disable=assignment-from-no-return
+        # let's assume the random state is already there (next class)
+        self.set_std_data(data + self.random_state.normal(size=data.shape), deterministic=False)  # type: ignore
```

Comment:
Dont' we want to change the `uid` in `mutate`?

Reply:
Nop, otherwise we'd need to update `uid` in `recombine` as well, then if we perform 1 mutation and 1 recombination, we lose track of the initial parent :s

Reply:
Basically the process is: spawn a child, then do whatever you want with it